### ISSUE TITLE
error C2079: '_' uses undefined struct 'fmt::v10::detail::type_is_unformattable_for<T,char>' 

### ISSUE THREAD

**Issue by hoyeziyu** [TheCherno/Hazel#683]
I wonder why? Thank you!
error C2079: '_' uses undefined struct 'fmt::v10::detail::type_is_unformattable_for<T,char>'
error C2338: Cannot format an argument. To make type T formattable provide a formatter specialization: https://fmt.dev/latest/api.html#udt

Log.h
```
#include <memory>
#include "Core.h"
#include "spdlog/spdlog.h"
#include "spdlog/fmt/ostr.h"

namespace Hazel {
	class HAZEL_API Log
	{
	public:
		static void Init();

		inline static std::shared_ptr<spdlog::logger>& GetCoreLogger() { return s_CoreLogger; }
		inline static std::shared_ptr<spdlog::logger>& GetClientLogger() { return s_ClientLogger; }
	private:
		static std::shared_ptr<spdlog::logger> s_CoreLogger;
		static std::shared_ptr<spdlog::logger> s_ClientLogger;
	};
}
```
Log.cpp
```
#include "Log.h"
#include "spdlog/sinks/stdout_color_sinks.h"


namespace Hazel {
	std::shared_ptr<spdlog::logger> Log::s_CoreLogger;
	std::shared_ptr<spdlog::logger> Log::s_ClientLogger;


	void Log::Init() {
		spdlog::set_pattern("%^[%T] %n：%v%$");
		s_CoreLogger = spdlog::stdout_color_mt("Hazel");
		s_CoreLogger->set_level(spdlog::level::trace);

		s_ClientLogger = spdlog::stdout_color_mt("App");
		s_ClientLogger->set_level(spdlog::level::trace);
	}
}
```
![image](https://github.com/user-attachments/assets/c0e73bc7-c704-49d6-8e77-abf94f217a2c)

**Comment by 323931** [comment]
bro，i have the same question, you fix it?

**Comment by VagueLobster** [comment]
> bro，i have the same question, you fix it?

There's a fix in Cherno's discord server in the Game-Engine-Series channel, but you might have to scroll up quite a bit 👍

**Comment by 323931** [comment]
> > bro，i have the same question, you fix it?
> 
> There's a fix in Cherno's discord server in the Game-Engine-Series channel, but you might have to scroll up quite a bit 👍

thank you bro,do you have a url? i do not know how to get the discord

**Comment by Kirdow** [comment]
> thank you bro,do you have a url? i do not know how to get the discord

It's in the readme

**Comment by 323931** [comment]
> > thank you bro,do you have a url? i do not know how to get the discord
> 
> It's in the readme

Thank you!!!!!!!!!

**Comment by hoho965191737** [comment]
No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you

**Comment by hoyeziyu** [comment]
> > bro，i have the same question, you fix it?
> 
> There's a fix in Cherno's discord server in the Game-Engine-Series channel, but you might have to scroll up quite a bit 👍

No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you

**Comment by hoyeziyu** [comment]
> bro，i have the same question, you fix it?

No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you

**Comment by 323931** [comment]
> > bro，i have the same question, you fix it?
> 
> No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you

bro，i fixed it，you can use e.ToString() instead of e
![微信图片_20240924111806](https://github.com/user-attachments/assets/1b396cd0-3a3f-4676-a9b1-619ad41d41fe)

**Comment by hoyeziyu** [comment]
> > > bro，i have the same question, you fix it?老哥，我也有同样的问题，你解决了吗？
> > 
> > 
> > No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you不，我还没解决。您找到解决问题的信息了吗？有没有具体的时间我们可以找到信息来解决问题？谢谢
> 
> bro，i fixed it，you can use e.ToString() instead of e兄弟，我修好了，你可以用 e.ToString() 代替 e ![微信图片_20240924111806](https://private-user-images.githubusercontent.com/153485319/370116982-1b396cd0-3a3f-4676-a9b1-619ad41d41fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNTE1NzAsIm5iZiI6MTcyNzE1MTI3MCwicGF0aCI6Ii8xNTM0ODUzMTkvMzcwMTE2OTgyLTFiMzk2Y2QwLTNhM2YtNDY3Ni1hOWIxLTYxOWFkNDFkNDFmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyNFQwNDE0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzQwNmFmMzZmNjU2NGZjZjY2YWU3MzBkZDE5ZTcyNzVkNjU0MmQzNTIzOTljZTk2MTQ2OTdlMWQ3ZDRkOTEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MlIpXJxms1517isYvRKsHZ1lV1hkVmc280K98VowV6Y)

Okay, thanks. I'll go back and try

**Comment by biigHandsomeGuy** [comment]
> > > bro，i have the same question, you fix it?
> > 
> > 
> > No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you
> 
> bro，i fixed it，you can use e.ToString() instead of e ![微信图片_20240924111806](https://private-user-images.githubusercontent.com/153485319/370116982-1b396cd0-3a3f-4676-a9b1-619ad41d41fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNTY0NzcsIm5iZiI6MTczNDM1NjE3NywicGF0aCI6Ii8xNTM0ODUzMTkvMzcwMTE2OTgyLTFiMzk2Y2QwLTNhM2YtNDY3Ni1hOWIxLTYxOWFkNDFkNDFmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQxMzM2MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWQ5ZTEyNDJiMGZhMTE4ZDRmYzE3ZmM2NDNkM2VjZjVkZWQyOWZhYmQzOTk3MWI2YTRhY2VhNTdiNGQ4ZjgzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.lf3lvuajNe_Be4P7dKhyOcrvnAMdH-7NokESNy8zbk8)

Thank's bro, it work

**Comment by jackpsmith-git** [comment]
Does anyone know how to solve this in the RegisterComponent() function in ScriptGlue.cpp ? None of the existing solutions seem applicable.

**Comment by hongyesuifeng** [comment]
> > > > 老哥，我也有同样的问题，你解决了吗？老哥，我也有同样的问题，你解决了吗？
> > > 
> > > 
> > > 不，我还没解决。您找到解决问题的信息了吗？有没有具体的时间我们可以找到信息来解决问题？谢谢不，我还没有解决。您找到解决问题的信息了吗？没有具体的时间我们可以找到解决问题的信息吗？谢谢
> > 
> > 
> > 兄弟，我修好了，你可以用e.ToString()代替e兄弟，我修好了，你可以用e.ToString()代替e![微信图片_20240924111806](https://private-user-images.githubusercontent.com/153485319/370116982-1b396cd0-3a3f-4676-a9b1-619ad41d41fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNTE1NzAsIm5iZiI6MTcyNzE1MTI3MCwicGF0aCI6Ii8xNTM0ODUzMTkvMzcwMTE2OTgyLTFiMzk2Y2QwLTNhM2YtNDY3Ni1hOWIxLTYxOWFkNDFkNDFmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyNFQwNDE0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzQwNmFmMzZmNjU2NGZjZjY2YWU3MzBkZDE5ZTcyNzVkNjU0MmQzNTIzOTljZTk2MTQ2OTdlMWQ3ZDRkOTEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MlIpXJxms1517isYvRKsHZ1lV1hkVmc280K98VowV6Y)
> 
> 好的，谢谢。我会回去试试

 thank you this help

**Comment by Xtremely-Doped-Punk** [comment]
> > > bro，i have the same question, you fix it?
> > 
> > 
> > No, I haven't solved it yet. Did you find the information that solved the problem? Is there a specific time when we can locate the information to solve the problem? Thank you
> 
> bro，i fixed it，you can use e.ToString() instead of e ![微信图片_20240924111806](https://github.com/user-attachments/assets/1b396cd0-3a3f-4676-a9b1-619ad41d41fe)

i dont think so, this defeats the purpose intended to, as in to pass the object as arg, and using the overloaded function for operater<<(Event e) definition, it must be able to change it to string automaticlly

**Comment by Xtremely-Doped-Punk** [comment]
here is the solution comment made in discord server. I am just forwarding it, hope it helps:

> this worked with older version, I guess:
```
inline std::ostream& operator<<(std::ostream& os, const Event& e)
{
    return os << e.ToString();
}
```

> But for newer version we need to use this instead:
```
inline std::string format_as(const Event& e) {
    return e.ToString();
}
```

i have tried this, and it works. hopefully we can close this issue :)
Refernce: https://fmt.dev/latest/api/#udt
> There are two ways to make a user-defined type formattable: providing a format_as function or specializing the formatter struct template. The format_as function should take an object of your type and return an object of a formattable type.

**Comment by UnpaintedKitchens** [comment]
hey, I have this same problem except with the types `glm::vec3`, `const std::filesystem::path`, `std::filesystem::path`, and `Hazel::UUID`. how would i create the format_as functions for these?

edit: nvm fixed it myself 👍

**Comment by Mavericks2019** [comment]
useing version 1.2.1 spdlog can solve this problem
